velox/exec/tests/ArrowStreamTest.cpp

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/exec/tests/utils/OperatorTestBase.h"
+#include "velox/exec/tests/utils/PlanBuilder.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::exec::test;
+
+// From Google gutil.
+#ifndef DISALLOW_COPY_AND_ASSIGN
+#define DISALLOW_COPY_AND_ASSIGN(TypeName) \
+  TypeName(const TypeName&) = delete;      \
+  void operator=(const TypeName&) = delete
+#endif
+
+class ArrowStreamTest : public OperatorTestBase {
+ protected:
+  /// Helper class for release callbacks.
+  class Helper {
+   public:
+    /// Query whether the C array stream is released.
+    static int ArrowArrayStreamIsReleased(
+        const struct ArrowArrayStream* stream) {
+      return stream->release == NULL;
+    }
+
+    /// Mark the C array stream released (for use in release callbacks).
+    static void ArrowArrayStreamMarkReleased(struct ArrowArrayStream* stream) {
+      stream->release = NULL;
+    }
+
+    /// Mark the C array released (for use in release callbacks).
+    static void ArrowArrayMarkReleased(struct ArrowArray* array) {
+      array->release = NULL;
+    }
+
+    /// Mock schema release.
+    static void mockSchemaRelease(ArrowSchema*) {}
+
+    /// Mock array release.
+    static void mockArrayRelease(ArrowArray*) {}
+
+    /// Mark array release callback as mock release.
+    static void ArrowArrayMockRelease(struct ArrowArray* array) {
+      array->release = mockArrayRelease;
+    }
+  };
+
+  /// A mock Arrow reader returns an Arrow array for each call of GetNext.
+  class ArrowReader {
+   public:
+    ArrowReader(
+        std::shared_ptr<memory::MemoryPool> pool,
+        const std::vector<RowVectorPtr>& vectors,
+        const TypePtr& type)
+        : pool_(pool), vectors_(vectors), type_(type) {}
+
+    int GetNext(struct ArrowArray* outArray) {
+      if (vectorIndex_ < vectors_.size()) {
+        exportToArrow(vectors_[vectorIndex_], *outArray, pool_.get());
+        vectorIndex_ += 1;
+        Helper::ArrowArrayMockRelease(outArray);
```

Comment:
Do we need these one-liner methods? I feel they make it harder to understand this code.

Reply:
Fixed cases like this.